mtags/src/main/scala-2/scala/meta/internal/mtags/OnDemandSymbolIndex.scala

```diff
@@ -51,90 +51,127 @@ final class OnDemandSymbolIndex(
 
   override def definition(symbol: Symbol): Option[SymbolDefinition] = {
-    try findSymbolDefinition(symbol, symbol)
+    try findSymbolDefinition(symbol, symbol).headOption
     catch onErrorOption
   }
 
-  override def addSourceDirectory(dir: AbsolutePath, dialect: Dialect): Unit =
-    tryRun {
-      if (sourceJars.addEntry(dir)) {
-        dir.listRecursive.foreach {
-          case source if source.isScala =>
-            try addSourceFile(source, Some(dir), dialect)
-            catch {
-              case NonFatal(e) => onError.lift(IndexError(source, e))
-            }
-          case _ =>
-        }
+  override def definitions(symbol: Symbol): List[SymbolDefinition] =
+    try findSymbolDefinition(symbol, symbol)
+    catch onError.andThen(_ => List.empty)
+
+  override def addSourceDirectory(
+      dir: AbsolutePath,
+      dialect: Dialect
+  ): List[(String, AbsolutePath)] =
+    tryRun(
+      List.empty, {
+        if (sourceJars.addEntry(dir)) {
+          dir.listRecursive.toList.flatMap {
+            case source if source.isScala =>
+              addSourceFile(source, Some(dir), dialect).map(sym =>
+                (sym, source)
+              )
+            case _ =>
+              List.empty
+          }
+        } else
+          List.empty
       }
-    }
+    )
 
   // Traverses all source files in the given jar file and records
   // all non-trivial toplevel Scala symbols.
-  override def addSourceJar(jar: AbsolutePath, dialect: Dialect): Unit =
-    tryRun {
-      try {
-        if (sourceJars.addEntry(jar)) {
-          FileIO.withJarFileSystem(jar, create = false) { root =>
-            root.listRecursive.foreach {
-              case source if source.isScala =>
-                try addSourceFile(source, None, dialect)
-                catch {
-                  case NonFatal(e) => onError.lift(IndexError(source, e))
-                }
-              case _ =>
+  override def addSourceJar(
+      jar: AbsolutePath,
+      dialect: Dialect
+  ): List[(String, AbsolutePath)] =
+    tryRun(
+      List.empty, {
+        try {
+          if (sourceJars.addEntry(jar)) {
+            FileIO.withJarFileSystem(jar, create = false) { root =>
+              root.listRecursive.toList.flatMap {
+                case source if source.isScala =>
+                  addSourceFile(source, None, dialect).map(sym => (sym, source))
+                case _ =>
+                  List.empty
+              }
             }
-          }
+          } else
+            List.empty
+        } catch {
+          case e: ZipError =>
+            onError(InvalidJarException(jar, e))
+            List.empty
         }
-      } catch {
-        case e: ZipError =>
-          onError(InvalidJarException(jar, e))
       }
-    }
+    )
 
   // Used to add cached toplevel symbols to index
-  def addSourceJarTopLevels(
+  def addIndexedSourceJar(
       jar: AbsolutePath,
-      loadTopLevels: () => TrieMap[String, AbsolutePath]
```

Comment:
I think this was done so that `loadTopLevels` is not calculated again if a jar was already added. Doesn't your change make it evaluate always? We could maybe make it `=> symbols: List[(String, AbsolutePath)]` and make it lazy val at invocation at the `None` case? It would serve the same purpose.

Reply:
It slightly changes behavior but jar-indexing definitely happens only once as it should
The difference in case if a jar was already added:
 - previously: absolutely nothing happens
 - now: for the first jar occurrence during the import it will request toplevels once from H2 and they will be ignored in this function.

I think it's how it should work.
Also, I've just found that we don't recreate index during build reloading that might lead to stale symbols so we can't rely on `DefinitionIndex` state.

Reply:
Makes sense! Let's leave it as is.